google/cloud/storage/testing/retry_tests.h

```diff
@@ -320,6 +315,232 @@ void PermanentFailureTest(Client& client,
+ * - Call the API.
+ * - Verify an error status is returned, with the right contents.
+ *
+ * This function implements these tests, saving us a lot of repetitive code.
+ *
+ * @tparam ReturnType The low-level RawClient return type (modulo StatusOr).
+ * @tparam F a formal parameter, the googlemock representation of the mocked
+ *     call.
+ *
+ * @param oncall The internal type returned by EXPECT_CALL(...).
+ * @param tested_operation a function wrapping the operation to be tested.
+ * @param api_name the name of the api
+ */
+template <typename ReturnType, typename F>
+void IdempotentFailuresStatusTest(
+    std::shared_ptr<testing::MockClient> const& mock,
+    ::testing::internal::TypedExpectation<F>& oncall,
+    std::function<Status(Client& client)> const& tested_operation,
+    char const* api_name) {
+  using canonical_errors::TransientError;
+  using ::testing::HasSubstr;
+  using ::testing::Return;
+  // A storage::Client with the strict idempotency policy, and with an
+  // easy-to-test retry policy.
+  Client client{std::shared_ptr<internal::RawClient>(mock),
+                StrictIdempotencyPolicy(), LimitedErrorCountRetryPolicy(2)};
+
+  // Expect exactly 3 calls before the retry policy is exhausted and an error
+  // status is returned.
+  oncall.WillOnce(Return(StatusOr<ReturnType>(TransientError())))
+      .WillOnce(Return(StatusOr<ReturnType>(TransientError())))
+      .WillOnce(Return(StatusOr<ReturnType>(TransientError())));
+
+  // Verify the right error status, with the right content, is returned when
+  // calling the operation.
+  Status status = tested_operation(client);
+  EXPECT_EQ(TransientError().status_code(), status.status_code());
+  EXPECT_THAT(status.error_message(), HasSubstr("Retry policy exhausted"));
+  EXPECT_THAT(status.error_message(), HasSubstr(api_name));
+}
+
+/**
+ * Test operations that are idempotent or not depending of their parameters.
+ *
+ * Some operations are idempotent when some parameters are set (preconditions).
+ * When the operation is idempotent, we want to verify that they are retried
+ * multiple times, when they are not, we want to retry them only once if the
+ * right policy is set.
+ *
+ * - Create a `storage::Client` with an easy-to-test retry policy.
+ * - Setup the mock to return the right number of transient failures.
+ * - Call the API.
+ * - Verify an error status is returned, with the right contents.
+ *
+ * This function implements these tests, saving us a lot of repetitive code.
+ *
+ * @tparam ReturnType The low-level RawClient return type (modulo StatusOr).
+ * @tparam F a formal parameter, the googlemock representation of the mocked
+ *     call.
+ *
+ * @param oncall The internal type returned by EXPECT_CALL(...).
+ * @param tested_operation a function wrapping the operation to be tested.
+ * @param api_name the name of the api
+ */
+template <typename ReturnType, typename F>
+void TooManyFailuresStatusTest(
```

Comment:
Is it intentional that this function is overloaded with the one above of the same name?

Reply:
Yes. It matches the `TooManyFailuresTest` earlier, which is also overloaded. We are not using both overloads yet because all the `*Notification()` functions are idempotent.